### DISCUSSION TITLE
Example for FC01 read coils

### DISCUSSION THREAD

**Discussion by Gamelauncher** [eModbus/eModbus#86] [Q&A]
Does someone have a short piece of code to generate a valid fc01 response? I've managed to solve all other responses but this one has me baffled a bit. It generates a Data Mismatch: response data length incorrect (single bit).

```
  if (addr < 8 && (addr + words) <= 8) {
    uint8_t returnData = 0;
    uint8_t returnBytes = (words + 7) / 8;
    response.add(request.getServerID(), request.getFunctionCode(), (uint8_t) returnBytes);    
    for (uint16_t i=0; i<words; ++i) {

      returnData << 1;
      returnData |= bitRead(outputStatusArr[addr+i], 0);

    }
    response.add( returnData );
  }

**Comment by Miq1** [comment]
AFAIK the coils need to be arranged LSB first, register-wise. So the lowest coil bit would be bit 0 in the first register, then coil 2 in bit 1 and so on up to coil 16 in bit 15. The next register would have coil 17 in bit 0, followed by coil 18 in bit 1 and further up to coil 32 in bit 15 etc.
The registers need to be MSB-first in the message.

So I would suggest having a ``uint16_t reg = 0;`` and adding the coils bit-shifted left for 0 to 15 positions one by one. Finally do an ``add(reg);`` to have it MSB first.

**Reply by Gamelauncher** [reply]
Found the problem:

https://www.fernhillsoftware.com/help/drivers/modbus/modbus-protocol.html

The count is in multiple's of 8 so it needs a uint8_t for the buffer. Ironed out the issue's, here is an example that "seems" to work ... not sure yet if I need to send the bytes in reverse order as well though. Most of the test kits interpret it differently in one way or another.

```
  addr--;

  if (addr < 266 && (addr + words) <= 266) {
    uint8_t returnData = 0;
    uint8_t currentByte = 0;
    uint8_t returnBytes = (words + 7) / 8;
    response.add(request.getServerID(), request.getFunctionCode(), (uint8_t) returnBytes);    
    for (uint16_t i=0; i<words; ++i) {

      returnData = returnData << 1;
      returnData = returnData | bitRead(outputStatusArr[addr+i], 0);
      
      if (currentByte == 7) {
        returnData = reverse(returnData);
        response.add( returnData );
        currentByte = 0;
        returnData = 0;
      } else {
        currentByte++;
      }
    }
    
    if (currentByte != 0) { // send the remainder of the bits
      returnData = reverse(returnData);
      response.add( returnData );
    }
  } else {
    response.setError(request.getServerID(), request.getFunctionCode(), ILLEGAL_DATA_ADDRESS);
  }

**Reply by Miq1** [reply]
Thank you for correcting and the code snippet. I am currently considering to add a ``CoilData`` data type to the lib, since the questions regarding coil handling are adding up. 

We will see.

**Reply by Gamelauncher** [reply]
In continuation ... is there a way to get the values of a FC15 / FC16 request? I've checked the message function but I don't see a way for a request.get multiple values :-)

**Reply by Miq1** [reply]
Without a loop? No. But that is not *that* complicated, is it? 😁

**Reply by Gamelauncher** [reply]
Found the loop, now need to find a piece of test software to be able actually send the command :-)

```
  for (auto& byte : request) {
    Serial.printf("%02X ", byte);
  }
  Serial.println();

**Reply by Miq1** [reply]
The code snippet is not really related - or is it? 

I rather had in mind something  alongside
```
for (uint16_t i = 0; i < numBytes; i++) {
  Offset = request.get(Offset, byteArray[i]) ;
}
```
